Merge V3 `@salesforce/retail-react-app` Features/Changes

bendvc opened the pull request:
# Description

This is a "long-lived" branch encapsulating all the changes and features we are adding to the next version (v3.0.0) of `@salesforce/retail-react-app`. 

- [ ] ⚠️  **Please merge with a regular merge (not squash)!**

# Rules 📏

1. Please keep track of any PRs merged into this branch so we have a single place to see a summary of changes  and feature additions. This is also helpful when creating release docs and change logs.
2. **Before cutting and merging your PRs from and into this branch, please ensure this branch is up-to-date with `develop`**. (This will save time for everyone 🙏🏻)
3. Before merging this this PR into `develop` lets double check that all the change logs and documentation is complete.

# PRs/Features/Changes Made

(See the updated [Changelog](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/v3/template-retail-react-app/packages/template-retail-react-app/CHANGELOG.md) from retail-react-app)


# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

bendvc committed:
Bump `template-retail-react-app` to 3.0.0-dev

bendvc committed:
Refactor components

bendvc committed:
Make mega menu fully lazy

bendvc committed:
Mobile working but rough

bendvc committed:
Reorganize component files

bendvc committed:
Refactor hoc's

bendvc committed:
File cleanup

bendvc committed:
Revert some changes to the nested accordion

bendvc committed:
make files exports consistent, rename hoc

bendvc committed:
Update CHANGELOG.md

bendvc committed:
Fix some tests

bendvc committed:
Fix test component import

bendvc committed:
Add tests for useCommerceSdkReact HOC

bendvc committed:
Fix broken menu after prop name change

bendvc committed:
Add proper sub-menus on product list page.

bendvc committed:
Use component class for placeholder prop, not a component element/instance

bendvc committed:
Fix box shadow not being visible

bendvc committed:
Generate translations

bendvc committed:
PR feedback

bendvc committed:
Merge branch 'develop' into v3/template-retail-react-app

bendvc committed:
Merge branch 'v3/template-retail-react-app' into breaking/lazy-load-categories

bendvc committed:
Fix tests

bendvc committed:
Merge pull request #1656 from SalesforceCommerceCloud/breaking/lazy-load-categories

[Retail React App V3] - Make Navigation Menus Lazy

alexvuong committed:
Merge branch 'develop' into v3/template-retail-react-app

bendvc committed:
Adapt menu after api changes

bendvc committed:
Correct styling for links without children

bendvc committed:
Add underline for menuless triggers

bendvc committed:
Revert commerce config

bendvc committed:
Merge pull request #1673 from SalesforceCommerceCloud/breaking/fix-menu-leaves

[Retail React App V3] - Lazy Navigation Menus Revision

alexvuong committed:
lazy basket creation

alexvuong committed:
lint

alexvuong committed:
remove mutations from useCurrentBasket

alexvuong committed:
reverted to have updateCustomerForBasket back

alexvuong committed:
Merge branch 'v3/template-retail-react-app' into v3-lazy-basket-creation

alexvuong committed:
reverted proxy config

alexvuong committed:
Merge pull request #1678 from SalesforceCommerceCloud/v3/retail-app-fix-revert-proxy

V3/retail app fix revert proxy

alexvuong committed:
Merge branch 'v3/template-retail-react-app' into v3-lazy-basket-creation

alexvuong committed:
Merge pull request #1677 from SalesforceCommerceCloud/v3-lazy-basket-creation

W-15040197: lazy basket creation

bendvc committed:
Merge branch 'develop' into v3/template-retail-react-app

bendvc committed:
Merge branch 'develop' into v3/template-retail-react-app

bendvc committed:
Re-lock mono repo

alexvuong commented:
@bendvc I noticed the filters on PLP are not open by default. STR
1. Run the Retail app
2. on home page, click on category nav
3. On PLP, none of the filter will not open on first load

bendvc committed:
Initial commit

- Attributes visible

bendvc committed:
Image changes now

bendvc committed:
Working with single selectable attribute

bendvc committed:
Update index.jsx

bendvc committed:
Move helper to utils

bendvc committed:
Merge branch 'develop' into bendvc/W-13974769

bendvc committed:
Merge branch 'develop' into v3/template-retail-react-app

bendvc committed:
Merge branch 'v3/product-tile-revamp' into bendvc/W-13974769

bendvc committed:
Adding comments

bendvc committed:
Expose props as constants

bendvc committed:
Always default to image defined on product if no image group is found

bendvc committed:
Only use circle swatches

bendvc committed:
Add tests

bendvc committed:
Refactor `getDecoratedVariationAttributes` and add tests

bendvc committed:
Bump main bundle size limit

bendvc committed:
Dynamic event handling for device type

bendvc committed:
Use useBreakpointValue to determine device type

bendvc committed:
Rework the swatch component

bendvc committed:
Render swatch as button when no href is supplied

bendvc committed:
Update index.jsx

bendvc committed:
Fix constant used for tile image on plp

bendvc committed:
Fix test and lint warnings

bendvc committed:
Bump bundle size again

bendvc committed:
Update packages/template-retail-react-app/app/components/product-tile/index.jsx

Co-authored-by: Vincent Marta <vmarta@salesforce.com>
Signed-off-by: Ben Chypak <bchypak@mobify.com>

bendvc committed:
Use module imports

bendvc committed:
Include image information for other pages where product tiles are used.

bendvc committed:
Add tests

bendvc committed:
Add explicit ariaLabel prop

alexvuong committed:
@W-15654904Showing strikethrough price on PLP (#1760)

* Display pricing for PLP and PDP

bendvc committed:
Update from base and update price on selected variation attribute

bendvc committed:
Lint!

bendvc committed:
fix test data after merge

bendvc committed:
Bump bundle size

alexvuong committed:
add changelog for product tile new pricing (#1792)

bendvc committed:
Don't use "hitType" to determine if variants exists

bendvc committed:
lint

bendvc committed:
Update packages/template-retail-react-app/app/components/product-tile/index.jsx

Signed-off-by: Ben Chypak <bchypak@mobify.com>

bendvc committed:
Merge pull request #1773 from SalesforceCommerceCloud/bendvc/W-13974769

@W-13974769 - Show Color Attribute on ProductTile 🀄️

unandyala committed:
@W-15714251: Show badges on product tile (#1791)

* Product Badges

bendvc committed:
Remove comment about updating code. It just works

kevinxh committed:
remove eslint disable comment

bendvc committed:
Merge pull request #1798 from SalesforceCommerceCloud/bendvc/W-15784798

W-15784798 - Finalize Lazy Navigation Changes

kevinxh committed:
Merge pull request #1799 from SalesforceCommerceCloud/minor-eslint-fix

Remove eslint disable comment

vmarta committed:
[Product Tile] Promotion callout message (@W-13974753@) (#1786)

* strikethrough for PLP

* add param for getProducts in useEinstein

* PR feedback

* Refactoring getDisplayPrice

* remove unused variables

* remove unvalid param

* take closest tiered price into account

* fix test

* put back recommendation section

* changing label

* fix bundle size

* fix tierPrices filter logic

* fix tierPrices filter logic

* fix tierPrices filter logic

* fix tierPrices filter logic

* fix lint

* add more cases for display price util

* add more cases for display price util

* fix logic in getDisplayPrice

* PR feedback

* PR feedback

* fix tests

* Apply formattedmessage to display price, refactor price info logic in getPriceData

* remove console

* guard the values

* fix tests

* adjust price display

* fix tests

* Promo callout: rough attempt at implementing it

* Allow customization at both page and component levels

* replace formatted string with jsx logic

* fix tests naming

* fiz translations

* Pr feedback

* fix template

* tweak proptype

* Consider when selecting a swatch

On swatch selection, the active variants can be more than 1. For example, the red colour is selected, but it may have several sizes.

* let bold font for PLP

* optimizing display price logic

* edit jsx

* PR feedback

* lint

* fix test

* build translations

* Let the DisplayPrice determine the variant with lowest price

* Convert to PromoCallout component instead

* PR feedback

* separate pricing components into its own files

* lint

* fix flaky test

* Rough refactoring

* Further refactoring and more flexible util method

* Make sure to render something (even empty div)

* Memoize the priceData

* PDP: render promo callout

* Add unit tests

* Add some test for the PDP

* findLowestPrice now also returns the applicable promotion object

* Add jsdoc comments

* Add more tests

* Consider swatch selection: first attempt

* Allow for null/undefined

* On swatch selection, show/hide the promo callout accordingly

* Memoize findLowestPrice call

* Simplify code block

* Move to outside the component

* Make sure to include 0 in the calculation

---------

Co-authored-by: Alex Vuong <alex.vuong@salesforce.com>

alexvuong committed:
W-14308527 Pricing display on cart, checkout, wishlist pages (#1796)

* price display on cart checkout and wishlist pages

alexvuong committed:
add changelog for PR1796 (#1807)

vmarta committed:
[Product Tile] Promo callout on the Product Details page (@W-15784291@) (#1804)

* PDP: more accurate price/promo when narrowing down to a variant

* Comment to clarify

* Workaround for API inconsistency

* Tweak logic

* Some refactoring for clarity

* Revert workaround

* Add tests

* Add more test

* Add tests for product set

* More test for product set

* Checking for `product.variants`

* Update CHANGELOG.md

* Make console.warn message to be more future-proof

alexvuong committed:
W-13974904 - Product tile revampe e2e tests (#1808)

* adjust e2e tests for product-tile revamp

unandyala committed:
@W-15752101: Tests for variant prices in PLP (#1817)

* Tests for variant prices in PLP

unandyala committed:
Fix Swatch Tests (#1819)

alexvuong committed:
W-12627197 -v3/template-retail-app Indicate live regions for dynamically changing content  (#1825)

* a11y live region for components

bendvc committed:
Fix accessibility tabbing into the size swatches

bendvc committed:
Merge pull request #1832 from SalesforceCommerceCloud/bug/fix-pdp-size-tabbing

[BUG][Accessibility] Fix Tab index Size Swatches on Product Detail Page

bendvc committed:
Merge branch 'develop' into v3/template-retail-react-app

bendvc committed:
Re-lock package.json files

vmarta committed:
E2E tests: revert url to the correct environment

vmarta committed:
Merge branch 'v3/template-retail-react-app' into v3/product-tile-revamp

alexvuong committed:
fix price display (#1853)

vmarta committed:
Merge pull request #1847 from SalesforceCommerceCloud/v3/product-tile-revamp

Merge v3/product tile revamp (@W-15945047@)

vmarta committed:
Update CHANGELOG.md

vmarta committed:
Update CHANGELOG.md

vmarta reviewed (dismissed):
Just to note: although this feature branch is for retail-react-app, we do update other packages/areas too:
- E2E tests
- commerce-sdk-react: we needed a new helper method for the lazy-basket-creation
- all packages: we re-generated the package lock files

alexvuong commented:
commerce-sdk-react: we needed a new helper method for the lazy-basket-creation => We did have a helper for the basket lazy creation. [Here](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/v3/template-retail-react-app/packages/commerce-sdk-react/src/hooks/ShopperBaskets/helpers.ts)

vmarta committed:
E2E: fix typo

vmarta committed:
Merge branch 'develop' into v3/template-retail-react-app

vmarta committed:
Update package-lock.json

vmarta committed:
Merge branch 'develop' into v3/template-retail-react-app

vmarta committed:
Merge branch 'develop' into v3/template-retail-react-app

vmarta committed:
Bump dev versions

It turns out that 3.6.0-dev is already published to npm 2 months ago. So for our tests to pass (specifically for when generating projects), we need to be using a version that is not published yet to npm.
